src/main/java/frc/robot/subsystems/drive/ModuleIOSpark.java

```diff
@@ -0,0 +1,237 @@
+// available in the root directory of this project.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+package frc.robot.subsystems.drive;
+
+import static frc.robot.Constants.DriveConstants.*;
+import static frc.robot.util.SparkUtil.*;
+
+import com.revrobotics.AbsoluteEncoder;
+import com.revrobotics.RelativeEncoder;
+import com.revrobotics.spark.ClosedLoopSlot;
+import com.revrobotics.spark.SparkBase;
+import com.revrobotics.spark.SparkBase.ControlType;
+import com.revrobotics.spark.SparkBase.PersistMode;
+import com.revrobotics.spark.SparkBase.ResetMode;
+import com.revrobotics.spark.SparkClosedLoopController;
+import com.revrobotics.spark.SparkClosedLoopController.ArbFFUnits;
+import com.revrobotics.spark.SparkFlex;
+import com.revrobotics.spark.SparkLowLevel.MotorType;
+import com.revrobotics.spark.SparkMax;
+import com.revrobotics.spark.config.ClosedLoopConfig.FeedbackSensor;
+import com.revrobotics.spark.config.SparkBaseConfig.IdleMode;
+import com.revrobotics.spark.config.SparkFlexConfig;
+import com.revrobotics.spark.config.SparkMaxConfig;
+import edu.wpi.first.math.MathUtil;
+import edu.wpi.first.math.filter.Debouncer;
+import edu.wpi.first.math.geometry.Rotation2d;
+import java.util.Queue;
+import java.util.function.DoubleSupplier;
+
+/**
+ * Module IO implementation for Spark Flex drive motor controller, Spark Max
+ * turn motor controller,
+ * and duty cycle absolute encoder.
+ */
+public class ModuleIOSpark implements ModuleIO {
+        private final Rotation2d zeroRotation;
+
+        // Hardware objects
+        private final SparkBase driveSpark;
+        private final SparkBase turnSpark;
+        private final RelativeEncoder driveEncoder;
+        private final AbsoluteEncoder turnEncoder;
+
+        // Closed loop controllers
+        private final SparkClosedLoopController driveController;
+        private final SparkClosedLoopController turnController;
+
+        // Queue inputs from odometry thread
+        private final Queue<Double> timestampQueue;
+        private final Queue<Double> drivePositionQueue;
+        private final Queue<Double> turnPositionQueue;
+
+        // Connection debouncers
+        private final Debouncer driveConnectedDebounce = new Debouncer(0.5);
+        private final Debouncer turnConnectedDebounce = new Debouncer(0.5);
+
+        public ModuleIOSpark(int module) {
+                zeroRotation = switch (module) {
+                        case 0 -> frontLeftZeroRotation;
+                        case 1 -> frontRightZeroRotation;
+                        case 2 -> backLeftZeroRotation;
+                        case 3 -> backRightZeroRotation;
+                        default -> new Rotation2d();
+                };
+                driveSpark = new SparkFlex(
```

Comment:
DOUBLE CHECK THIS PLEASE. 
All of our old code used SparkMax for both drive and turn.